linuxkm/lkcapi_glue.c

```diff
@@ -0,0 +1,2537 @@
+        goto test_gcm_end;
+    }
+
+    req = aead_request_alloc(tfm, GFP_KERNEL);
+    if (IS_ERR(req)) {
+        pr_err("error: allocating AES aead request %s failed: %ld\n",
+               WOLFKM_AESCBC_DRIVER, PTR_ERR(req));
+        goto test_gcm_end;
+    }
+
+    src = kmalloc(sizeof(struct scatterlist) * 2, GFP_KERNEL);
+    dst = kmalloc(sizeof(struct scatterlist) * 2, GFP_KERNEL);
+
+    if (IS_ERR(src) || IS_ERR(dst)) {
+        pr_err("error: kmalloc src or dst failed: %ld, %ld\n",
+               PTR_ERR(src), PTR_ERR(dst));
+        goto test_gcm_end;
+    }
+
+    sg_init_table(src, 2);
+    sg_set_buf(src, assoc2, sizeof(assoc));
+    sg_set_buf(&src[1], dec2, sizeof(vector));
+
+    sg_init_table(dst, 2);
+    sg_set_buf(dst, assoc2, sizeof(assoc));
+    sg_set_buf(&dst[1], enc2, decryptLen);
+
+    aead_request_set_callback(req, 0, NULL, NULL);
+    aead_request_set_ad(req, sizeof(assoc));
+    aead_request_set_crypt(req, src, dst, sizeof(vector), iv);
+
+    ret = crypto_aead_encrypt(req);
+
+    if (ret) {
+        pr_err("error: crypto_aead_encrypt returned: %d\n", ret);
+        goto test_gcm_end;
+    }
+
+    ret = XMEMCMP(enc, enc2, sizeof(vector));
+    if (ret) {
+        pr_err("error: enc and enc2 do not match: %d\n", ret);
+        goto test_gcm_end;
+    }
+
+    ret = XMEMCMP(authTag, enc2 + encryptLen, sizeof(authTag));
+    if (ret) {
+        pr_err("error: authTags do not match: %d\n", ret);
+        goto test_gcm_end;
+    }
+
+    /* Now decrypt crypto request. Reverse src and dst. */
+    memset(dec2, 0, decryptLen);
+    aead_request_set_ad(req, sizeof(assoc));
+    aead_request_set_crypt(req, dst, src, decryptLen, iv);
+
+    ret = crypto_aead_decrypt(req);
+
+    if (ret) {
+        pr_err("error: crypto_aead_decrypt returned: %d\n", ret);
+        goto test_gcm_end;
+    }
+
+    ret = XMEMCMP(dec, dec2, sizeof(vector));
+    if (ret) {
+        pr_err("error: dec and dec2 do not match: %d\n", ret);
+        goto test_gcm_end;
+    }
+
+test_gcm_end:
+    if (req) { aead_request_free(req); req = NULL; }
+    if (tfm) { crypto_free_aead(tfm); tfm = NULL; }
+
+    if (src) { kfree(src); src = NULL; }
+    if (dst) { kfree(dst); dst = NULL; }
+
+    if (dec2) { kfree(dec2); dec2 = NULL; }
+    if (enc2) { kfree(enc2); enc2 = NULL; }
+
+    if (assoc2) { kfree(assoc2); assoc2 = NULL; }
+    if (iv) { kfree(iv); iv = NULL; }
+
+    return ret;
+}
+
+#endif /* HAVE_AESGCM &&
+        * (LINUXKM_LKCAPI_REGISTER_ALL || LINUXKM_LKCAPI_REGISTER_AESGCM) &&
+        * (! (WOLFSSL_AESNI && WC_AES_C_DYNAMIC_FALLBACK))
+        */
+
+#if defined(WOLFSSL_AES_XTS) && \
+    (defined(LINUXKM_LKCAPI_REGISTER_ALL) || defined(LINUXKM_LKCAPI_REGISTER_AESXTS))
+
+#ifndef HEAP_HINT
+#define HEAP_HINT NULL
+#endif
+
+#ifndef ERROR_OUT
+#define ERROR_OUT(err, eLabel) do { ret = (err); goto eLabel; } while (0)
+#endif
+
+#ifndef WC_TEST_RET_ENC_EC
+#define WC_TEST_RET_ENC_EC(ret) (ret)
+#endif
+
+#ifndef WC_TEST_RET_ENC_NC
```

Comment:
Don't want macros for testing in the code.

Reply:
fixed.

Reply:
I'm still seeing ERROR_OUT and HEAP_HINT and WC_USE_DEVID and devId.
These aren't necessary in this code.

Reply:
OK OK I'll scrub out the last of that.  I was trying to avoid these routines drifting too far from test.c, whence they came, but I think that ship has already sailed, and now it shall sail yet further.

Reply:
fixed.